### PR TITLE
functions/imagemagick: system tests + Node 8 upgrade

### PR DESCRIPTION
**Pull request by ace-n**

### CODE REVIEW

**Comment by fhinkel** [comment]
@ace-n The imagemagick tests are failing. Ping me when it's ready for another review

**Comment by ace-n** [comment]
@fhinkel tests are passing now, this should be good to review + merge. Thanks for your patience!

File: functions/imagemagick/index.js [hunk 1]
```diff
@@ -30,115 +30,94 @@ const {BLURRED_BUCKET_NAME} = process.env;
...
-    });
+  try {
+    const [result] = await client.safeSearchDetection(filePath);
+    const detections = result.safeSearchAnnotation || {};
+
+    if (
+      detections.adult === 'VERY_LIKELY' ||
+      detections.violence === 'VERY_LIKELY'
+    ) {
+      console.log(`The image ${file.name} has been detected as inappropriate.`);
```
Comment (grayside) [inline functions/imagemagick/index.js]: Concise, up-front context, and active tense makes for more easily scanned and understood log messages.

```suggestion
      console.log(`Detected image ${file.name} as inappropriate.`);
```

File: functions/imagemagick/index.js [hunk 2]
```diff
@@ -30,115 +30,94 @@ const {BLURRED_BUCKET_NAME} = process.env;
...
+      return blurImage(file, BLURRED_BUCKET_NAME);
+    } else {
+      console.log(`The image ${file.name} has been detected as OK.`);
```
Comment (grayside) [inline functions/imagemagick/index.js]:
```suggestion
      console.log(`Detected image ${file.name} as OK.`);
```

File: functions/imagemagick/index.js [hunk 3]
```diff
@@ -30,115 +30,94 @@ const {BLURRED_BUCKET_NAME} = process.env;
...
+    }
+  } catch (err) {
+    console.error(`Failed to analyze ${file.name}.`, err);
```
Comment (grayside) [inline functions/imagemagick/index.js]:
```suggestion
    console.error(`Failed analysis for ${file.name}:`, err);
```

File: functions/imagemagick/index.js [hunk 4]
```diff
@@ -30,115 +30,94 @@ const {BLURRED_BUCKET_NAME} = process.env;
...
-        `Blurred image has been uploaded to: gs://${blurredBucketName}/${file.name}`
-      );
-
-      // Delete the temporary file.
-      return new Promise((resolve, reject) => {
-        fs.unlink(tempLocalPath, err => {
+  try {
+    await file.download({destination: tempLocalPath});
+
+    console.log(`Image ${file.name} has been downloaded to ${tempLocalPath}.`);
```
Comment (grayside) [inline functions/imagemagick/index.js]:
```suggestion
    console.log(`Downloaded image ${file.name} to ${tempLocalPath}.`);
```

File: functions/imagemagick/index.js [hunk 5]
```diff
@@ -30,115 +30,94 @@ const {BLURRED_BUCKET_NAME} = process.env;
...
+  } catch (err) {
+    console.error('Failed to download file.', err);
```
Comment (grayside) [inline functions/imagemagick/index.js]:
```suggestion
    console.error(`Failed download: ${file.name}:`, err);
```

File: functions/imagemagick/index.js [hunk 6]
```diff
@@ -30,115 +30,94 @@ const {BLURRED_BUCKET_NAME} = process.env;
...
+
+  try {
+    // Blur the image using ImageMagick.
+    await new Promise((resolve, reject) => {
+      gm(tempLocalPath)
+        .blur(0, 16)
+        .write(tempLocalPath, (err, stdout) => {
           if (err) {
-            reject(err);
+            console.error('Failed to blur image.', err);
```
Comment (grayside) [inline functions/imagemagick/index.js]: This log statement is inside the promise resolution. Save this for the catch statement.

```suggestion
```

File: functions/imagemagick/index.js [hunk 7]
```diff
@@ -30,115 +30,94 @@ const {BLURRED_BUCKET_NAME} = process.env;
...
+            return reject(err);
           } else {
-            resolve();
+            resolve(stdout);
           }
         });
-      });
     });
-}
+    console.log(`Image ${file.name} has been blurred.`);
```
Comment (grayside) [inline functions/imagemagick/index.js]:
```suggestion
    console.log(`Blurred image ${file.name}.`);
```

File: functions/imagemagick/index.js [hunk 8]
```diff
@@ -30,115 +30,94 @@ const {BLURRED_BUCKET_NAME} = process.env;
...
+  } catch (err) {
+    console.error('Failed to blur image.', err);
```
Comment (grayside) [inline functions/imagemagick/index.js]:
```suggestion
    console.error(`Failed image transform: blur {file.name}:`, err);
```

File: functions/imagemagick/index.js [hunk 9]
```diff
@@ -30,115 +30,94 @@ const {BLURRED_BUCKET_NAME} = process.env;
...
+  // Upload result to a different bucket, to avoid re-triggering this function.
+  // You can also re-upload it to the same bucket + tell your Cloud Function to
+  // ignore files marked as blurred (e.g. those with a "blurred" prefix)
+  const blurredBucket = storage.bucket(blurredBucketName);
+
+  // Upload the Blurred image back into the bucket.
+  try {
+    await blurredBucket.upload(tempLocalPath, {destination: file.name});
+    console.log(
+      `Blurred image has been uploaded to: gs://${blurredBucketName}/${file.name}`
```
Comment (grayside) [inline functions/imagemagick/index.js]:
```suggestion
      `Blurred image uploaded to: gs://${blurredBucketName}/${file.name}`
```

File: functions/imagemagick/index.js [hunk 10]
```diff
@@ -30,115 +30,94 @@ const {BLURRED_BUCKET_NAME} = process.env;
...
+    );
+  } catch (err) {
+    console.error('Failed to upload blurred image.', err);
```
Comment (grayside) [inline functions/imagemagick/index.js]:
```suggestion
    console.error(`Failed upload: gs://${blurredBucketName}/${file.name}:`, err);
```

File: functions/imagemagick/index.js [hunk 11]
```diff
@@ -30,115 +30,94 @@ const {BLURRED_BUCKET_NAME} = process.env;
...
+    return Promise.reject(err);
+  }
+
+  // Delete the temporary file.
+  return new Promise((resolve, reject) => {
```
Comment (grayside) [inline functions/imagemagick/index.js]: I think it's cleaner practice to use [util.promisify()](https://nodejs.org/api/util.html#util_util_promisify_original)

File: functions/imagemagick/test/index.test.js [hunk 12]
```diff
@@ -15,169 +15,115 @@
...
 
-it('blurOffensiveImages blurs unblurred images (Node 8 syntax)', async () => {
-  const sample = getSample(defaultFileName);
+    const {stdout} = await stopFF(ffProc);
 
-  await sample.program.blurOffensiveImages({
-    bucket: bucketName,
-    name: defaultFileName,
+    assert.ok(
+      stdout.includes(`The image ${safeFileName} has been detected as OK.`)
```
Comment (grayside) [inline functions/imagemagick/test/index.test.js]:
```suggestion
      stdout.includes(`Detected image ${safeFileName} as OK.`)
```

File: functions/imagemagick/test/index.test.js [hunk 13]
```diff
@@ -15,169 +15,115 @@
...
+          name: offensiveFileName,
         },
-      ])
-    ),
+      },
+    });
+
+    const {stdout} = await stopFF(ffProc);
+
+    assert.ok(stdout.includes(`Image ${offensiveFileName} has been blurred.`));
```
Comment (grayside) [inline functions/imagemagick/test/index.test.js]:
```suggestion
    assert.ok(stdout.includes(`Blurred image ${safeFileName}.`));
```

File: functions/imagemagick/test/index.test.js [hunk 14]
```diff
@@ -15,169 +15,115 @@
...
+    assert.ok(
+      stdout.includes(
+        `Blurred image has been uploaded to: gs://${BLURRED_BUCKET_NAME}/${offensiveFileName}`
```
Comment (grayside) [inline functions/imagemagick/test/index.test.js]:
```suggestion
        `Blurred image uploaded to:
gs://${BLURRED_BUCKET_NAME}/${offensiveFileName}`
```

File: functions/imagemagick/index.js (hunk 8 above)
Comment (ace-n) [reply functions/imagemagick/index.js]: AFAICT our samples generally log _either_ the `err` object _or_ the relevant parameter (i.e. `file.name`), but _not both_.

Should we keep this as-is, or replace `err` with `file.name`?

**Comment by grant** [comment]
Going to remove myself as a reviewer as there already are active reviewers.
Thanks for the FYI.

**Comment by fhinkel** [comment]
Linter is complaining and a few other required tests don't pass. Can you have another look?

**Comment by fhinkel** [comment]
The `imagemagick` tests are failing again :(

File: functions/imagemagick/index.js (hunk 11 above)
Comment (ace-n) [reply functions/imagemagick/index.js]: Turns out `promisify` doesn't work here. (No idea why - perhaps it doesn't have the right callback signature? 🤔)

**Pull request merged**